dashboard/app/models/course.rb

```diff
@@ -250,39 +250,52 @@ def scripts_for_user(user)
   # course script (or the default course script itself) by evaluating these
   # rules in order:
   #
-  # 1. If the user is a student, and the student has progress in (or has been
-  # assigned to) the script associated with one of the default or alternate
-  # course scripts, return that script. If there is more than one such script,
-  # return the one that was assigned/progressed/updated most recently.
+  # 1. If the user is a teacher, and they have a course experiment enabled,
+  # show the corresponding alternate course script.
   #
-  # 2. Return the first alternate course script for which the teacher (either
-  # the current user, or the teacher of the student's most recently-joined
-  # section) has the corresponding experiment enabled, if one exists.
+  # 2. If the user is in a section assigned to this course: show an alternate
+  # course script if any section's teacher is in a corresponding course
+  # experiment, otherwise show the default course script.
   #
-  # 3. otherwise, return the default course script.
+  # 3. If the user is a student and has progress in an alternate course script,
+  # show the alternate course script.
+  #
+  # 4. Otherwise, show the default course script.
   #
   # @param user [User|nil]
   # @param default_course_script [CourseScript]
   # @return [CourseScript]
   def select_course_script(user, default_course_script)
+    return default_course_script unless user
+
     alternates = alternate_course_scripts.where(default_script: default_course_script.script).all
 
-    if user.try(:student?)
-      course_scripts = alternates + [default_course_script]
+    if user.teacher?
+      alternates.each do |cs|
+        return cs if SingleUserExperiment.enabled?(user: user, experiment_name: cs.experiment_name)
+      end
+    end
 
-      # include hidden scripts when iterating over user scripts.
-      user.user_scripts.each do |us|
-        course_script = course_scripts.find {|cs| cs.script == us.script}
-        return course_script if course_script
+    course_sections = user.sections_as_student.where(course: self)
+    unless course_sections.empty?
+      course_sections.each do |section|
+        alternates.each do |cs|
+          return cs if SingleUserExperiment.enabled?(user: section.teacher, experiment_name: cs.experiment_name)
+        end
```

Comment:
If we're in multiple sections that are assigned to this course, and one of those sections has experiment A and one has experiment B, we're going to end up with whichever course_section we see first, right?

(I'm not sure to what extent this is a scenario we need to care about)

Reply:
Yep, that's correct. So whichever alternate course script we list first in our config will take precedence over the other, in a way. I think this is fine and will probably never happen, and the main thing we care about here is consistent behavior and not blowing up if a user gets into this state.

Reply:
Oh wait, you said it right and I said it wrong. It will be based on which section we see first, not which alternate course script is listed first in the config. I will switch the nesting order here, since that might make things a little bit more predictable.